File: EPSILON/TOOLS/Altova/Altova/XMLSpy2006/spl/cs/AltovaText/TableLike/FLF/FLFParser.cs
[call GenerateFileHeader("FLFParser.cs")]
using System;

namespace Altova.TextParser.TableLike.FLF
{
	/// <summary>
	/// Encapsulates parsing a fixed length fields format.
	/// </summary>
	public class FLFParser : RecordBasedParser
	{
		#region Implementation Detail:
		FLFFormat mFormat = null;
		int mOffset = 0;
		int mMaxOffset = 0;
		string mBuffer = string.Empty;

		void ParseFields(string\[\] fields, int fieldcount)
		{
			if (mFormat.AssumeRecordDelimitersPresent)
			{
				int i=0;
				while (i < fieldcount)
				{
					int count = Header\[i\].Length;
					int recSepPos = mBuffer.IndexOf('\\r', mOffset, count);
					
					if (recSepPos != -1 && recSepPos-mOffset <= count)
					{
						count = recSepPos-mOffset;
						fields\[i\] = mBuffer.Substring(mOffset, count).TrimEnd(mFormat.FillCharacter);
						mOffset +=count;
						i++;
						break;
					}
					else
					{
						fields\[i\] = mBuffer.Substring(mOffset, count).TrimEnd(mFormat.FillCharacter);
						mOffset += count;
						i++;
					}
				}
				if (i < fieldcount)
					for (int j=i; j< fieldcount; j++)
						fields\[j\] = "";
			}
			else
			{
				for (int i = 0; i < fieldcount; ++i)
				{
					fields\[i\] = mBuffer.Substring(mOffset, Header\[i\].Length).TrimEnd(mFormat.FillCharacter);
					mOffset += Header\[i\].Length;
				}
			}
		}
		
		void ParseRecord()
		{
			string\[\] fields = new string\[Header.Count\];

			if (mFormat.AssumeRecordDelimitersPresent)
			{
				int maxfieldindex = Header.Count - 1;
				this.ParseFields(fields, maxfieldindex);
				int start = mOffset;
				this.MoveToNextRecordDelimiter();
				int len = Math.Min(mOffset - start, Header\[maxfieldindex\].Length);
				fields\[maxfieldindex\] = mBuffer.Substring(start, len).TrimEnd(mFormat.FillCharacter);
				this.SwallowTillNextRecord();
			}
			else this.ParseFields(fields, Header.Count);

			if (null != OnRecordFound) OnRecordFound(new Record(fields));
		}
		void MoveToNextNonRecordDelimiter()
		{
			while ((mOffset < mBuffer.Length) && (mFormat.IsRecordDelimiter(mBuffer\[mOffset\])))
				++mOffset;
		}
		void MoveToNextRecordDelimiter()
		{
			while ((mOffset < mBuffer.Length) && (!mFormat.IsRecordDelimiter(mBuffer\[mOffset\])))
				++mOffset;
		}
		void SwallowTillNextRecord()
		{
			this.MoveToNextRecordDelimiter();
			this.MoveToNextNonRecordDelimiter();
		}
		#endregion
		#region Public Interface:
		/// <summary>
		/// Constructs an instance of this class.
		/// </summary>
		/// <param name="format">the format information to be used by this instance</param>
		/// <param name="header">the header information to be used by this instance</param>
		public FLFParser(FLFFormat format, Header header) : base(header)
		{
			mFormat = format;
		}
		/// <summary>
		/// Gets the format specification the parser is using.
		/// </summary>
		public FLFFormat Format
		{
			get
			{
				return mFormat;
			}
		}
		#endregion
		#region Implementing RecordBasedParser:
		/// <summary>
		/// Parses a buffer as fixed length fields format.
		/// </summary>
		/// <param name="buffer">the buffer to parse</param>
		/// <returns>the number of records found</returns>
		public override int Parse(string buffer)
		{
			int result = 0;
			mBuffer = buffer;
			mMaxOffset = mBuffer.Length - Header.RecordLength;
			mOffset = 0;
			
			if (mFormat.AssumeRecordDelimitersPresent)
			{
				int minLength=0;
				for (int i = 0; i<Header.Count-1; i++)
					minLength += Header\[i\].Length;
				if (mBuffer.Length-2 < minLength)
					throw new FLFParserException("The input buffer is not long enough for even just one record.", mBuffer.Length);
			}
			else
				if (0 > mMaxOffset)
					throw new FLFParserException("The input buffer is not long enough for even just one record.", mBuffer.Length);
			do
			{
				this.ParseRecord();
				++result;
			}
			while (mOffset < mBuffer.Length);
			return result;
		}
		#endregion
	}
}

File: EPSILON/TOOLS/Altova/Altova/XMLSpy2006/spl/cpp/AltovaText/FLFParser.cpp
[call GenerateFileHeader("FLFParser.cpp")]
#include "stdafx.h"
#include "FLFParser.h"
#include "Helpers.h"

namespace altova
{
namespace text
{
namespace tablelike
{
namespace flf
{

// CParserException:
CParserException::CParserException(tstring message, size_t offset)
:	m_Offset(offset)
{
#ifdef _UNICODE
	std::wstringstream ss;
#else
	std::stringstream ss;
#endif
	ss << message << _T(" at offset #") << offset;
	CMappingException::SetMessage(ss.str());
}
size_t CParserException::GetOffset() const
{
	return m_Offset;
}

// ---------------------------------------------------------------------------------------

CParser::CParser(const flf::CFormat& format, const CHeader& header)
:	m_Format(format)
,	m_Header(header)
,	m_Offset(0)
,	m_MaxOffset(0)
,	m_Buffer(_T(""))
{
}

int CParser::Parse(const tstring& buffer)
{
	int result = 0;
	m_Buffer = buffer;
	m_MaxOffset = m_Buffer.size() - GetRecordLength();
	m_Offset = 0;
	
	if (m_Format.GetAssumeRecordDelimiters())
	{
		size_t minLength =0;
		for (size_t i = 0; i < m_Header.GetColumns().size()-1; ++i)
			minLength += m_Header.GetColumns()\[i\]->GetLength();
		if (m_Buffer.size()-2 < minLength)
			throw CParserException(_T("The input buffer is not long enough for even just one record."), m_Buffer.length());
	}
	else
		if (0>m_MaxOffset)
			throw CParserException(_T("The input buffer is not long enough for even just one record."), m_Buffer.length());
	
	do
	{
		ParseRecord();
		++result;
	}
	while (m_Offset<m_Buffer.length());
	return result;
}

int CParser::GetRecordLength()
{
	int result = 0;
	const TColumnSpecificationArray& cols = m_Header.GetColumns();
	TColumnSpecificationArray::const_iterator it, itEnd = cols.end();
	for (it = cols.begin(); it != itEnd; ++it)
		result += (*it)->GetLength();
	return result;
}

size_t CParser::ExtractFieldValueFromBuffer(int bufferstartoffset, TStringList& fields, size_t fieldindex)
{
	size_t len = m_Header.GetColumns()\[fieldindex\]->GetLength();
	tstring value = m_Buffer.substr(bufferstartoffset, len);
	Trim(value);
	fields.push_back(new tstring(value));
	return len;
}

void CParser::ParseFields(TStringList& fields, size_t fieldcount)
{
	if (m_Format.GetAssumeRecordDelimiters())
	{
		size_t i=0;
		while (i < fieldcount)
		{
			int count = m_Header.GetColumns()\[i\]->GetLength();
			tstring value = m_Buffer.substr(m_Offset, count);
			size_t recSepPos = value.find(_T('\\r'), 0);
			if (recSepPos != tstring::npos)
			{
				value.resize(recSepPos);
				Trim(value);
				fields.push_back(new tstring(value));
				m_Offset += recSepPos;
				i++;
				break;
			}
			else
			{
				Trim(value);
				fields.push_back(new tstring(value));
				m_Offset += count;
				i++;
			}
		}
		if (i<fieldcount)
			for (size_t j = i; j < fieldcount; j++)
				fields\[j\] = new tstring(_T(""));
	}
	else
	{
		for (size_t i = 0; i < fieldcount; ++i)
		{
			size_t len = m_Header.GetColumns()\[i\]->GetLength();
			tstring value = m_Buffer.substr(m_Offset, len);
			Trim(value);
			fields.push_back(new tstring(value));
			m_Offset += len;
		}
	}
}

void CParser::ParseRecord()
{
	TStringList fields;
	const TColumnSpecificationArray& cols= m_Header.GetColumns();
	if (m_Format.GetAssumeRecordDelimiters())
	{
		size_t maxfieldindex = cols.size() - 1;
		ParseFields(fields, maxfieldindex);
		size_t start = m_Offset;
		MoveToNextRecordDelimiter();
		size_t len = min(m_Offset-start, cols\[maxfieldindex\]->GetLength());
		tstring value = m_Buffer.substr(start, len);
		Trim(value);
		fields.push_back(new tstring(value));

		SwallowTillNextRecord();
	}
	else 
	{
		ParseFields(fields, cols.size());
	}
	CRecordBasedParser::NotifyObserverAboutRecordFound(fields);
	DeleteContainedPointers(fields);
}

void CParser::MoveToNextNonRecordDelimiter()
{
	while ((m_Offset<m_Buffer.size()) && (m_Format.IsRecordDelimiter(m_Buffer\[m_Offset\])))
		++m_Offset;
}

void CParser::MoveToNextRecordDelimiter()
{
	while ((m_Offset<m_Buffer.size()) && (!m_Format.IsRecordDelimiter(m_Buffer\[m_Offset\])))
		++m_Offset;
}

void CParser::SwallowTillNextRecord()
{
	MoveToNextRecordDelimiter();
	MoveToNextNonRecordDelimiter();
}

void CParser::Trim(tstring& rhs)
{
	if(rhs.empty()) return;
	TCHAR fillcharacter = m_Format.GetFillCharacter();
	size_t size = rhs.size();
	int end = size - 1;
	while ((end>-1) && (fillcharacter==rhs\[end\]))
		--end;
	rhs = rhs.substr(0, end+1);
}

// ---------------------------------------------------------------------------------------

} // namespace flf
} // namespace tablelike
} // namespace text
} // namespace altova
